kernel/fsgs.c

```diff
@@ -12,61 +12,46 @@
 
 void myst_set_fsbase(void* p)
 {
-    if (__options.have_syscall_instruction)
+    if (__myst_kernel_args.have_fsgsbase_instructions)
     {
-        const long n = SYS_arch_prctl;
-        myst_syscall2(n, ARCH_SET_FS, (long)p);
+        __asm__ volatile("wrfsbase %0" ::"r"(p));
+    }
+    else if (__options.have_syscall_instruction)
+    {
+        myst_syscall2(SYS_arch_prctl, ARCH_SET_FS, (long)p);
     }
     else
     {
-        __asm__ volatile("wrfsbase %0" ::"r"(p));
+        myst_panic("unsupported");
     }
 }
 
 void* myst_get_fsbase(void)
 {
-    if (__options.have_syscall_instruction)
-    {
-        const long n = SYS_arch_prctl;
-        void* p;
-        myst_syscall2(n, ARCH_GET_FS, (long)&p);
-        return p;
-    }
-    else
-    {
-        void* p;
-        __asm__ volatile("mov %%fs:0, %0" : "=r"(p));
-        return p;
-    }
+    void* p;
+    __asm__ volatile("mov %%fs:0, %0" : "=r"(p));
```

Comment:
To be accurate, this need to be `rdfsbase` instruction.

Reply:
Thank you Anand, I considered that too but we would need to emulate RDFSBASE in OE too, which we don't. We are relying on the fact that in all of our scenarios, we expect to find a structure first field contains the address of the thread pointer itself.